### PR TITLE
ref(js): General cleanup to AutoComplete

### PR DESCRIPTION
**Pull request by evanpurkhiser**

### CODE REVIEW

File: static/app/components/autoComplete.tsx [hunk 1]
```diff
@@ -51,21 +51,34 @@ type GetInputOutput<E extends HTMLInputElement> = GetInputArgs<E> &
     value?: string;
   };
 
-export type GetItemArgs<T> = {
+type GetItemArgs<T> = {
```
Comment (evanpurkhiser) [inline static/app/components/autoComplete.tsx:54]: Wasn't being used anywhere

File: static/app/components/autoComplete.tsx [hunk 2]
```diff
@@ -157,14 +179,13 @@ class AutoComplete<T extends Item> extends React.Component<Props<T>, State<T>> {
...
-    this.setState({
-      highlightedIndex: this.props.defaultHighlightedIndex || 0,
-    });
+    this.setState({highlightedIndex: this.props.defaultHighlightedIndex ?? 0});
   }
 
-  handleInputChange =
-    <E extends HTMLInputElement>({onChange}: Pick<GetInputArgs<E>, 'onChange'>) =>
-    (e: React.ChangeEvent<E>) => {
+  makeHandleInputChange<E extends HTMLInputElement>(
```
Comment (evanpurkhiser) [inline static/app/components/autoComplete.tsx:185]: These are all factory functions, renamed them as such.

They also don't need to be instance properties, they can just be class methods

File: static/app/components/autoComplete.tsx [hunk 3]
```diff
@@ -205,7 +227,7 @@ class AutoComplete<T extends Item> extends React.Component<Props<T>, State<T>> {
   }
 
   // Dropdown detected click outside, we should close
-  async handleClickOutside() {
+  handleClickOutside = async () => {
```
Comment (evanpurkhiser) [inline static/app/components/autoComplete.tsx:230]: Removing a `.bind(this)`

File: static/app/components/autoComplete.tsx [hunk 4]
```diff
@@ -311,7 +335,7 @@ class AutoComplete<T extends Item> extends React.Component<Props<T>, State<T>> {
 
     // when this component is in virtualized mode, only a subset of items will be passed
     // down, making the array length inaccurate. instead we manually pass the length as itemCount
-    const listSize = this.itemCount || this.items.size;
+    const listSize = this.itemCount ?? this.items.size;
```
Comment (evanpurkhiser) [inline static/app/components/autoComplete.tsx:338]: A bit more correct, we don't want `itemCount` 0 defaulting to `item.size` I believe

File: static/app/components/autoComplete.tsx [hunk 5]
```diff
@@ -231,7 +254,7 @@ class AutoComplete<T extends Item> extends React.Component<Props<T>, State<T>> {
       if (hasHighlightedItem && (canSelectWithEnter || canSelectWithTab)) {
         const item = this.items.get(this.state.highlightedIndex);
 
-        if (!item.disabled) {
+        if (item && !item.disabled) {
```
Comment (pythonsen) [inline static/app/components/autoComplete.tsx:257]: `item && !item.disabled ` can be  `!item?.disabled `

Comment (evanpurkhiser) [reply static/app/components/autoComplete.tsx:257]: It could if I didn’t want to also tell typescript that item is set for the enclosed function call :)

**Review by matejminar — APPROVED** [review]

**Pull request merged**